File: src/campus_desk/perms.py
"""权限位定义与查询（M4 设计 v3 §2 + M6 RBAC 三表化）。

M6 起运行时以 DB 为准：get_role_perms / effective_perms_from_db 查
roles/permissions/role_permissions 三表；下方 ROLE_PERMS / GRANTABLE_PERMS
降级为"种子数据源 + 兜底"（与 db/seed.py 三表种子同源、与前端 constants/perms.js
同源；未知角色/未种子库回退用）。users.permissions 列存附加权限位（逗号分隔），
最终权限 = 角色默认(查库) ∪ 附加位。
"""

from __future__ import annotations

from sqlalchemy import select

from campus_desk.db.models import Permission, RolePermission

# 角色默认权限位（M6 降级为兜底/种子源，运行时查 role_permissions 表）
ROLE_PERMS: dict[str, list[str]] = {
    "student": ["chat"],
    "cs_staff": ["chat", "cs_workbench"],
    "admin": ["chat", "cs_workbench", "kb_review", "view_stats", "user_mgmt", "view_logs"],
}

# 可被 admin 授予的附加权限位（M6 降级为兜底/种子源，运行时查 permissions 表）
GRANTABLE_PERMS: list[str] = [
    "cs_workbench",
    "kb_review",
    "view_stats",
    "user_mgmt",
    "view_logs",
]


def effective_perms(role: str, extra: str = "") -> list[str]:
    """最终权限 = 角色默认 ∪ 附加位（纯函数兜底，extra: 逗号分隔字符串）。"""
    base = ROLE_PERMS.get(role, ["chat"])
    extras = [s.strip() for s in (extra or "").split(",") if s.strip()]
    return list(dict.fromkeys([*base, *extras]))


def get_role_perms(session, role: str) -> list[str]:
    """查库取角色默认权限（role_permissions join permissions，按权限 id 排序）。

    角色在 roles 表无任何关联行（含未知角色）→ 回退 ROLE_PERMS 兜底，保证不锁死。
    """
    rows = (
        session.execute(
            select(Permission.id)
            .join(RolePermission, RolePermission.permission_id == Permission.id)
            .where(RolePermission.role_id == role)
            .order_by(Permission.id)
        )
        .scalars()
        .all()
    )
    return list(rows) or ROLE_PERMS.get(role, ["chat"])


def effective_perms_from_db(session, role: str, extra: str = "") -> list[str]:
    """最终权限 = 角色默认(查库) ∪ 附加位（login 时调用，结果写 JWT claims）。"""
    base = get_role_perms(session, role)
    extras = [s.strip() for s in (extra or "").split(",") if s.strip()]
    return list(dict.fromkeys([*base, *extras]))

File: frontend/src/constants/perms.js
// M4 权限体系（设计文档 §2）：角色默认权限 ∪ 附加权限位 = 最终权限
// M6-ZJUT 起运行时以 DB 为准：登录响应 user.permissions 已是角色权限(role_permissions 表) ∪
// 附加位的最终并集；下方 ROLE_PERMS / GRANTABLE_PERMS 仅作兜底（种子源），
// 用户管理页的勾选项已改为查 /api/admin/roles、/api/admin/permissions 接口。

export const PERM = {
  CHAT: 'chat',
  CS_WORKBENCH: 'cs_workbench',
  KB_REVIEW: 'kb_review',
  VIEW_STATS: 'view_stats',
  USER_MGMT: 'user_mgmt',
  VIEW_LOGS: 'view_logs'
}

// 角色默认权限位（与后端 require_perm 同源）
export const ROLE_PERMS = {
  student: [PERM.CHAT],
  cs_staff: [PERM.CHAT, PERM.CS_WORKBENCH],
  admin: [
    PERM.CHAT,
    PERM.CS_WORKBENCH,
    PERM.KB_REVIEW,
    PERM.VIEW_STATS,
    PERM.USER_MGMT,
    PERM.VIEW_LOGS
  ]
}

// 可被 admin 授予的附加权限位（用户管理页勾选用）
export const GRANTABLE_PERMS = [
  { key: PERM.CS_WORKBENCH, label: '客服工作台（接待标记）' },
  { key: PERM.KB_REVIEW, label: '知识库审查补入' },
  { key: PERM.VIEW_STATS, label: '数据看板' },
  { key: PERM.USER_MGMT, label: '用户管理' },
  { key: PERM.VIEW_LOGS, label: '日志管理' }
]

// 侧栏角色菜单（按最终权限过滤）
export const PERM_MENUS = [
  { path: '/cs', title: '客服工作台', icon: 'Headset', perm: PERM.CS_WORKBENCH },
  { path: '/admin', title: '知识库管理', icon: 'Collection', perm: PERM.KB_REVIEW },
  { path: '/stats', title: '数据看板', icon: 'DataLine', perm: PERM.VIEW_STATS },
  { path: '/users', title: '用户管理', icon: 'UserFilled', perm: PERM.USER_MGMT },
  { path: '/logs', title: '日志管理', icon: 'Document', perm: PERM.VIEW_LOGS }
]

// 最终权限 = 角色默认 ∪ 附加位
// extra 兼容两种输入：后端登录响应的 permissions 是数组；本地可传逗号字符串
export function effectivePerms(role, extra = '') {
  const base = ROLE_PERMS[role] || [PERM.CHAT]
  const extras = (Array.isArray(extra) ? extra : String(extra || '').split(','))
    .map((s) => String(s).trim())
    .filter(Boolean)
  return [...new Set([...base, ...extras])]
}

export function hasPerm(role, extra, perm) {
  return effectivePerms(role, extra).includes(perm)
}

// 当前登录用户（localStorage 非响应式，调用方需自行加响应式依赖）
export function currentUser() {
  try {
    return JSON.parse(localStorage.getItem('cd_user') || '{}')
  } catch {
    return {}
  }
}
